### PR TITLE
dx | 1201 | get and remove rate limit commands

### PR DESCRIPTION
**Pull request by harshithad0703**
- get rate limit command
- remove rate limit command
- version bump to be made

### CODE REVIEW

**Review by shafeeqfalcon — APPROVED** [review]

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import { cliux, configHandler, isAuthenticated } from '@contentstack/cli-utilities';
+import { Command } from '@contentstack/cli-command';
+import { handleErrorMsg } from '../../../utils/interactive';
+import { RateLimitConfig } from '../../../interfaces';
+
+export default class RateLimitGetCommand extends Command {
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/get/rate-limit.ts:6]: @harshithad0703 , description is missing

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+  static examples = ['$ csdx config:get:rate-limit'];
+
+  async run() {
+    try {
+      if (!isAuthenticated()) {
+        const err = { errorMessage: 'You are not logged in. Please login with command $ csdx auth:login' };
+        handleErrorMsg(err);
+      }
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/get/rate-limit.ts:15]: @harshithad0703 , authentication isn't required for this command.

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+          minWidth: 20,
+        },
+        Limit: {
+          minWidth: 20,
+        },
+        'Bulk Limit': {
+          minWidth: 20,
+        },
+        'Download Limit': {
+          minWidth: 20,
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/get/rate-limit.ts:43]: @harshithad0703 , Please remove it as we don't need it as the CS plan has no download limit.

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      const rateLimitData: RateLimitConfig = rateLimits.rateLimit || {};
+
+      const tableData = Object.entries(rateLimitData).map(([org, limits]) => ({
+        Org: org === 'default' ? 'default' : org,
+        'Get Limit': limits.getLimit ? `${limits.getLimit.value}(${limits.getLimit.utilize}%)` : '0',
+        Limit: limits.limit ? `${limits.limit.value}(${limits.limit.utilize}%)` : '0',
+        'Bulk Limit': limits.bulkLimit ? `${limits.bulkLimit.value}(${limits.bulkLimit.utilize}%)` : '0',
+        'Download Limit': limits.downloadLimit
+          ? `${limits.downloadLimit.value}(${limits.downloadLimit.utilize}%)`
+          : '0',
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/get/rate-limit.ts:26]: @harshithad0703 , Please remove it as we don't need it as the CS plan has no download limit.

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  static examples = ['$ csdx config:get:rate-limit'];
+
+  async run() {
+    try {
+      if (!isAuthenticated()) {
+        const err = { errorMessage: 'You are not logged in. Please login with command $ csdx auth:login' };
+        handleErrorMsg(err);
+      }
+      const rateLimits = configHandler.get('rateLimits') || {};
+      const rateLimitData: RateLimitConfig = rateLimits.rateLimit || {};
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/get/rate-limit.ts:17]: @harshithad0703 , pls check the structure properly. code should be like this

    const rateLimit = configHandler.get('rateLimit') || {};
    const formatLimit = (limit) => limit ? `${limit.value}(${limit.utilize}%)` : '0';

    const tableData = Object.entries(rateLimitData).map(([org, limits]: [string, RateLimitConfig]) => ({
    Org: org === 'default' ? 'default' : org,
    'Get Limit': formatLimit(limits.getLimit),
    Limit: formatLimit(limits.limit),
    'Bulk Limit': formatLimit(limits.bulkLimit),
}));

File: packages/contentstack-config/src/commands/config/remove/rate-limit.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import { cliux, configHandler, FlagInput, flags, isAuthenticated } from '@contentstack/cli-utilities';
+import { Command } from '@contentstack/cli-command';
+import { askOrgID, handleErrorMsg } from '../../../utils/interactive';
+
+export default class RateLimitRemoveCommand extends Command {
+
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/remove/rate-limit.ts:6]: @harshithad0703 , description is missing

File: packages/contentstack-config/src/commands/config/remove/rate-limit.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  };
+  static examples = [
+    '$ csdx config:remove:rate-limit --org <<org_uid>>'
+  ];
+  async run() {
+    try {
+      if (!isAuthenticated()) {
+        const err = { errorMessage: 'You are not logged in. Please login with command $ csdx auth:login' };
+        handleErrorMsg(err);
+      }
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/remove/rate-limit.ts:20]: @harshithad0703 , authentication isn't required

File: packages/contentstack-config/src/commands/config/remove/rate-limit.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      let { org } = flags;
+      if (!org) {
+        org = await askOrgID();
+      }
+      const rateLimits = configHandler.get('rateLimits') || {};
+
+      if (!rateLimits.rateLimit[org]) {
+        cliux.print(`No rate limit found for the organization UID: ${org}`, { color: 'red' });
+        return;
+      }
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/remove/rate-limit.ts:31]: @harshithad0703 , pls check the structure properly. 

      const rateLimit = configHandler.get('rateLimit') || {};

      if (!rateLimit[org]) {
        cliux.print(`No rate limit found for the organization UID: ${org}`, { color: 'red' });
        return;
      }

File: packages/contentstack-config/src/commands/config/remove/rate-limit.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+      delete rateLimits.rateLimit[org];
+
+      configHandler.set('rateLimits', rateLimits);
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/remove/rate-limit.ts:35]: @harshithad0703 , you can directly delete the config using 
configHandler.delete(`rateLimit.${org}`)

File: packages/contentstack-config/src/commands/config/remove/rate-limit.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+      cliux.print(`Rate limit entry for organization UID ${org} has been removed.`, { color: 'green' });
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/remove/rate-limit.ts:36]: @harshithad0703 , pls check all text with doc team.

File: packages/contentstack-config/src/utils/interactive.ts [hunk 11]
```diff
@@ -96,3 +96,23 @@ export async function askEarlyAccessHeaderAlias(): Promise<string> {
...
+  if (err?.errorMessage) {
+    cliux.print(`Error: ${err.errorMessage}`, { color: 'red' });
+  } else if (err?.message) {
+    cliux.print(`Error: ${err.message}`, { color: 'red' });
+  } else {
+    console.log(err);
+    cliux.print(`Error: ${messageHandler.parse('CLI_CONFIG_API_FAILED')}`, { color: 'red' });
+  }
+  process.exit(1);
+}
```
Comment (aman19K) [inline packages/contentstack-config/src/utils/interactive.ts:110]: @harshithad0703 , don't think it's required.

**Review by aman19K — CHANGES_REQUESTED** [review]

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts (hunk 1 above)
Comment (harshithad0703) [reply packages/contentstack-config/src/commands/config/get/rate-limit.ts:5]: added

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts (hunk 2 above)
Comment (harshithad0703) [reply packages/contentstack-config/src/commands/config/get/rate-limit.ts]: removed authentication here

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts (hunk 3 above)
Comment (harshithad0703) [reply packages/contentstack-config/src/commands/config/get/rate-limit.ts]: done

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts (hunk 4 above)
Comment (harshithad0703) [reply packages/contentstack-config/src/commands/config/get/rate-limit.ts:21]: done

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts (hunk 5 above)
Comment (harshithad0703) [reply packages/contentstack-config/src/commands/config/get/rate-limit.ts]: updated the structure

File: packages/contentstack-config/src/commands/config/remove/rate-limit.ts (hunk 7 above)
Comment (harshithad0703) [reply packages/contentstack-config/src/commands/config/remove/rate-limit.ts]: done

File: packages/contentstack-config/src/utils/interactive.ts (hunk 11 above)
Comment (harshithad0703) [reply packages/contentstack-config/src/utils/interactive.ts]: removed this as its not required

File: packages/contentstack-config/src/commands/config/remove/rate-limit.ts (hunk 8 above)
Comment (harshithad0703) [reply packages/contentstack-config/src/commands/config/remove/rate-limit.ts]: structure updated

File: packages/contentstack-config/src/commands/config/remove/rate-limit.ts (hunk 10 above)
Comment (harshithad0703) [reply packages/contentstack-config/src/commands/config/remove/rate-limit.ts:31]: will check with doc team

File: packages/contentstack-config/src/commands/config/remove/rate-limit.ts (hunk 9 above)
Comment (harshithad0703) [reply packages/contentstack-config/src/commands/config/remove/rate-limit.ts]: updated

File: packages/contentstack-config/src/commands/config/remove/rate-limit.ts (hunk 6 above)
Comment (harshithad0703) [reply packages/contentstack-config/src/commands/config/remove/rate-limit.ts:7]: added description

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+export default class RateLimitGetCommand extends Command {
+  static description: string = messageHandler.parse('Get rate-limit of organizations');
+  static examples = ['$ csdx config:get:rate-limit'];
+
+  async run() {
+    try {
+      const rateLimit = configHandler.get('rateLimit') || {};
+      const rateLimitData: RateLimitConfig = rateLimit || {};
+
+      const tableData = Object.entries(rateLimitData).map(([org, limits]) => ({
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/get/rate-limit.ts:14]: @harshithad0703 , pls check the shared example
https://github.com/contentstack/cli/pull/1543#discussion_r1741906689

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
+import { cliux, configHandler, messageHandler } from '@contentstack/cli-utilities';
+import { Command } from '@contentstack/cli-command';
+import { RateLimitConfig } from '../../../interfaces';
+
+export default class RateLimitGetCommand extends Command {
+  static description: string = messageHandler.parse('Get rate-limit of organizations');
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/get/rate-limit.ts:6]: @harshithad0703 , [Minor] messageHandler isn't required.

File: packages/contentstack-config/src/commands/config/remove/rate-limit.ts [hunk 14]
```diff
@@ -0,0 +1,33 @@
+import { cliux, configHandler, FlagInput, flags, messageHandler } from '@contentstack/cli-utilities';
+import { Command } from '@contentstack/cli-command';
+import { askOrgID } from '../../../utils/interactive';
+
+export default class RateLimitRemoveCommand extends Command {
+  static description: string = messageHandler.parse('Remove rate-limit of the organization');
```
Comment (aman19K) [inline packages/contentstack-config/src/commands/config/remove/rate-limit.ts:6]: @harshithad0703 , [Minor] messageHandler isn't required.

Comment (harshithad0703) [reply packages/contentstack-config/src/commands/config/remove/rate-limit.ts]: done

File: packages/contentstack-config/src/commands/config/get/rate-limit.ts (hunk 13 above)
Comment (harshithad0703) [reply packages/contentstack-config/src/commands/config/get/rate-limit.ts]: done

**Review by aman19K — APPROVED** [review]

**Pull request merged**